### PR TITLE
feat: create `@call` operator

### LINKED ISSUE #846: Create a `@call` operator
The `@call` operator will allow our users to call the resolver by name that is specified in the root Query and Mutation. Here is an example of how it would work.

```graphql
schema {
  query: Query
}

type Query {
  posts: [Post] @http(path: "/path")
  user(id: Int): User @http(path: "/user/{{args.id}}")
}

type Post {
  id: Int
  title: String
  userId: Int
  user: User @call(query: "user", args: {id: "{{value.userId}}"})
}
```

The `@call` will simply call the resolver that's set in `Query.user` field with the arguments specified via `args`.

This feature gives us the ability to define functions using the Tailcall DSL, which means we can add some form of reusability and modularity.

**Technical Requirements**
- The function can only be called if the input and output type match exactly. In the above example  `Query::user` returns a `User` and `Post::user` also returns a `User` so it's compatible with its return type. Similarly, we can see that the args are also exactly the same.

### PR DESCRIPTION
**Pull request by ologbonowiwi**
**Issue Reference(s):**  
Fixes #846 
/claim #846

**Build & Testing:**

- [x] I ran `cargo test` successfully.
- [x] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [x] I have added relevant unit & integration tests.
- [x] I have updated the [documentation] accordingly.
- [x] I have performed a self-review of my code.
- [x] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcallhq.github.io/tree/develop/docs

### CODE REVIEW

**Comment by tusharmath** [comment]
@ologbonowiwi I want to come back to this. Can you resolve conflicts once.

**Comment by ologbonowiwi** [comment]
@tusharmath conflicts solved on this one

File: src/config/expr.rs [hunk 1]
```diff
@@ -39,6 +39,10 @@ pub enum ExprBody {
     #[serde(rename = "graphQL")]
     GraphQL(GraphQL),
 
+    /// Reuses a resolver pre-defined on type `Query`
```
Comment (ologbonowiwi) [inline src/config/expr.rs:42]:
```suggestion
    /// Reuses a resolver pre-defined on `Query` or `Mutation`
```

File: examples/jsonplaceholder.graphql [hunk 2]
```diff
@@ -24,5 +24,5 @@ type Post {
   userId: Int!
   title: String!
   body: String!
-  user: User @http(path: "/users/{{value.userId}}")
+  user: User @call(query: "user", args: {id: "{{value.userId}}"})
```
Comment (tusharmath) [inline examples/jsonplaceholder.graphql:27]: Let's add another example config for using `call`

File: src/blueprint/mustache.rs [hunk 3]
```diff
@@ -101,14 +101,14 @@ impl<'a> MustachePartsValidator<'a> {
 }
 
 impl FieldDefinition {
-    pub fn validate_field(&self, type_of: &config::Type, config: &Config) -> Valid<(), String> {
+    pub fn validate_field(self, type_of: &config::Type, config: &Config) -> Valid<Self, String> {
```
Comment (tusharmath) [inline src/blueprint/mustache.rs:104]: Why are we making this change?

Comment (ologbonowiwi) [reply src/blueprint/mustache.rs:104]: Or open another PR to it

File: examples/jsonplaceholder.graphql (hunk 2 above)
Comment (tusharmath) [reply examples/jsonplaceholder.graphql:27]: Also as side note, in this case `user` doesn't actually accept `args`

File: tests/execution/call-graphql-datasource.md [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+schema
+  @server(port: 8000, graphiql: true, hostname: "0.0.0.0")
+  @upstream(baseURL: "http://jsonplaceholder.typicode.com", httpCache: true) {
+  query: Query
+}
+
+type Query {
+  posts: [Post] @http(path: "/posts")
+  user(id: Int!): User
+    @graphQL(baseURL: "http://upstream/graphql", name: "user", args: [{key: "id", value: "{{args.id}}"}])
```
Comment (tusharmath) [inline tests/execution/call-graphql-datasource.md:15]:
```suggestion
  user(id: Int!): User @http(path: "/users/{{args.id}}")
```
Let's use a simpler example here.

File: tests/execution/call-mutation.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+type Query {
+  firstUser: User @http(method: "GET", path: "/users/1")
+}
+
+type User {
+  id: Int
+  name: String
+  post: Post
+    @call(mutation: "insertPost", args: {input: "{\"body\":\"user-body\",\"title\":\"user-title\",\"userId\":1}"})
```
Comment (tusharmath) [inline tests/execution/call-mutation.md:36]: Writing the input as string encoded JSON isn't the most convinient. I think we should be able to pass the value using the graphQL syntax as follows:
```graphql
@call(mutation: "insertPost", args: {input: {body: "user-body", title: "user-title", userId: 1}})
```

File: tests/execution/call-graphql-datasource.md (hunk 4 above)
Comment (ologbonowiwi) [reply tests/execution/call-graphql-datasource.md:15]: But in this case, we'd not be testing call with `@graphql`

File: src/blueprint/operators/call.rs [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+                return Valid::succeed(b_field);
+            };
+
+            compile_call(field, config, call, operation_type)
+                .map(|resolver| b_field.resolver(Some(resolver)))
+        },
+    )
+}
+
+pub fn compile_call(
```
Comment (tusharmath) [inline src/blueprint/operators/call.rs:31]: I think this implementation can be massively simplified. We need to invent a new expression inside `Context`

```rust
pub enum Context {
    Value,
    Path(Vec<String>),
    
    // A new operator
    CallWith {
        args: HashMap<String, Vec<String>>,
        expression: Box<Expression>
    } 
}
```

Whenever we see a `call` config, we should use it's `args` field to modify the current args in context and then call the expression with the new context. That way we won't need to make changes here, when we add a new IO type.

File: src/blueprint/mustache.rs (hunk 3 above)
Comment (tusharmath) [reply src/blueprint/mustache.rs:104]: It takes a reference and validates. That's a good design for a validator. Let's revert this please.

File: src/blueprint/operators/const_field.rs [hunk 7]
```diff
@@ -35,23 +35,28 @@ pub fn compile_const(inputs: CompileConst) -> Valid<Expression, String> {
...
-            Ok(gql) => {
-                let validation = if validate {
-                    validate_data_with_schema(config_module, field, gql)
-                } else {
-                    Valid::succeed(())
-                };
-                validation.map(|_| Literal(data.to_owned()))
+    Valid::from(
+        DynamicValue::try_from(&value.clone()).map_err(|e| ValidationError::new(e.to_string())),
+    )
```
Comment (ologbonowiwi) [inline src/blueprint/operators/const_field.rs:40]: The change was necessary as this code was repeated across multiple places.

File: src/blueprint/operators/call.rs (hunk 6 above)
Comment (ologbonowiwi) [reply src/blueprint/operators/call.rs:33]: Mentioned on Discord but updating it here also: tried to implement this but did not succeed. I do have a WIP on [feat/#846/create-call-operator-via-resolver-with-arg](https://github.com/ologbonowiwi/tailcall/tree/feat/%23846/create-call-operator-via-resolver-with-arg), though.

File: src/blueprint/mustache.rs (hunk 3 above)
Comment (ologbonowiwi) [reply src/blueprint/mustache.rs]: Reverted.

File: tests/execution/test-call-operator-errors.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  id: Int!
+}
+
+type Post {
+  userId: Int!
+  withoutResolver: User @call(query: "userWithoutResolver", args: {id: "{{value.userId}}"})
+  withoutOperator: User @call(args: {id: "{{value.userId}}"})
+  urlMismatchHttp: User @call(query: "user", args: {})
+  argumentMismatchGraphQL: User @call(query: "userWithGraphQLResolver", args: {})
+  headersMismatchGraphQL: User @call(query: "userWithGraphQLResolver", args: {})
```
Comment (tusharmath) [inline tests/execution/test-call-operator-errors.md:30]:
```suggestion
  headersMismatchGraphQL: User @call(query: "userWithGraphQLResolver")
```
Args should be optional.

Comment (ologbonowiwi) [reply tests/execution/test-call-operator-errors.md]: > Args should be optional.

Changed. The `args` didn't had a default value, but I implemented it on [b199a05](https://github.com/tailcallhq/tailcall/pull/1064/commits/b199a05a64b910cf96a45a15edf3c4a2f6eab519)

> Error when no args are supplied

This is already being tested. The snapshot for this error is on [tests/snapshots/execution_spec__test-call-operator-errors.md_errors.snap](https://github.com/ologbonowiwi/tailcall/blob/feat/%23846/create-call-operator/tests/snapshots/execution_spec__test-call-operator-errors.md_errors.snap)

Comment (ologbonowiwi) [reply tests/execution/test-call-operator-errors.md]: > Changed. The args didn't had a default value, but I implemented it on [b199a05](https://github.com/tailcallhq/tailcall/pull/1064/commits/b199a05a64b910cf96a45a15edf3c4a2f6eab519)

Re-reading now, I'm not sure if this was the expected solution. Let me know if so. It made sense to make it optional as sometimes no arg is needed, and if it's the case, the user shouldn't be required to declare an "empty" `args` field.

File: src/blueprint/operators/call.rs (hunk 6 above)
Comment (tusharmath) [reply src/blueprint/operators/call.rs:33]: Adding a todo here — https://github.com/tailcallhq/tailcall/issues/1285

**Comment by tusharmath** [comment]
Thanks @ologbonowiwi!

Looking forward to it's documentation on the website :)
/tip 100$

**Review by tusharmath — APPROVED** [review]

**Pull request merged**